data-prepper-core/src/main/java/org/opensearch/dataprepper/core/pipeline/server/DynamicPipelineUpdateUtil.java

```diff
@@ -0,0 +1,130 @@
+                                                  final PipelinesDataFlowModel targetDataPipelineDataFlowMode) {
+
+        checkIfSamePipelineExists(currentPipelineDataFlowMode, targetDataPipelineDataFlowMode);
+        Set<String> singleThreadedProcessorNames = scanForSingleThreadAnnotatedProcessorPlugins();
+        singleThreadedProcessorNames.addAll(excludedProcessorsFromDynamicUpdate);
+
+        for (String pipelineName : currentPipelineDataFlowMode.getPipelines().keySet()) {
+            PipelineModel currentPipelineModel = currentPipelineDataFlowMode.getPipelines().get(pipelineName);
+            PipelineModel targetPipelineModel = targetDataPipelineDataFlowMode.getPipelines().get(pipelineName);
+
+            // Check if source configuration remains unchanged
+            PluginModel currentSource = currentPipelineModel.getSource();
+            PluginModel targetSource = targetPipelineModel.getSource();
+            if (!currentSource.equals(targetSource)) {
+                throw new DynamicPipelineConfigUpdateException(
+                        "Source configuration cannot be modified in pipeline: " + pipelineName);
+            }
+
+            // Check if sinks configuration remains unchanged
+            List<SinkModel> currentSinks = currentPipelineModel.getSinks();
+            List<SinkModel> targetSinks = targetPipelineModel.getSinks();
+            if (!currentSinks.equals(targetSinks)) {
+                throw new DynamicPipelineConfigUpdateException(
+                        "Sinks configuration cannot be modified in pipeline: " + pipelineName);
+            }
+
+            List<PluginModel> currentProcessors = currentPipelineModel.getProcessors();
+            List<PluginModel> targetProcessors = targetPipelineModel.getProcessors();
+
+            currentProcessors = currentProcessors == null ? List.of() : currentProcessors;
+            targetProcessors = targetProcessors == null ? List.of() : targetProcessors;
+
+            // Collect single-threaded processors in current and target
+            Set<String> currentSingleThreaded = currentProcessors.stream()
+                    .map(PluginModel::getPluginName)
+                    .filter(singleThreadedProcessorNames::contains)
+                    .collect(Collectors.toSet());
+
+            Set<String> targetSingleThreaded = targetProcessors.stream()
+                    .map(PluginModel::getPluginName)
+                    .filter(singleThreadedProcessorNames::contains)
+                    .collect(Collectors.toSet());
+
+            // Only throw if target adds new single-threaded processors or modifies existing ones
+            for (String targetProcessor : targetSingleThreaded) {
+                if (!currentSingleThreaded.contains(targetProcessor)) {
+                    throw new DynamicPipelineConfigUpdateException(
+                            "Cannot add new single-threaded processor: " + targetProcessor);
```

Comment:
Why do we care about this? Is it because of the state?

If so, this is insufficient. The `aggregate` processor retains state, but it is done in a thread-safe way.

Also, `grok` is currently labeled as `@SingleThread` because of a small code gap.

Maybe we should have an `@Stateful` annotation to indicate that state is kept?